features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/pinned/banner/PinnedMessagesBannerPresenter.kt

```diff
@@ -17,40 +17,145 @@
 import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.matrix.api.room.MatrixRoom
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.persistentListOf
+import kotlinx.collections.immutable.toImmutableList
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
 import javax.inject.Inject
+import kotlin.time.Duration.Companion.milliseconds
+
+class PinnedMessagesBannerPresenter @Inject constructor(
+    private val room: MatrixRoom,
+    private val itemFactory: PinnedMessagesBannerItemFactory,
+    private val isFeatureEnabled: IsPinnedMessagesFeatureEnabled,
+    private val networkMonitor: NetworkMonitor,
+) : Presenter<PinnedMessagesBannerState> {
+    private val pinnedItems = mutableStateOf<ImmutableList<PinnedMessagesBannerItem>>(persistentListOf())
 
-class PinnedMessagesBannerPresenter @Inject constructor() : Presenter<PinnedMessagesBannerState> {
     @Composable
     override fun present(): PinnedMessagesBannerState {
-        var pinnedMessageCount by remember {
-            mutableIntStateOf(0)
-        }
-        var currentPinnedMessageIndex by rememberSaveable {
-            mutableIntStateOf(0)
-        }
+        val isFeatureEnabled = isFeatureEnabled()
+        val knownPinnedMessagesCount by remember {
+            room.roomInfoFlow.map { roomInfo -> roomInfo.pinnedEventIds.size }
+        }.collectAsState(initial = 0)
+
+        var hasTimelineFailedToLoad by rememberSaveable { mutableStateOf(false) }
+        var currentPinnedMessageIndex by rememberSaveable { mutableIntStateOf(-1) }
+
+        PinnedMessagesBannerItemsEffect(
+            isFeatureEnabled = isFeatureEnabled,
+            onItemsChange = { newItems ->
+                val pinnedMessageCount = newItems.size
+                if (currentPinnedMessageIndex >= pinnedMessageCount || currentPinnedMessageIndex < 0) {
+                    currentPinnedMessageIndex = pinnedMessageCount - 1
+                }
+                pinnedItems.value = newItems
+            },
+            onTimelineFail = { hasTimelineFailed ->
+                hasTimelineFailedToLoad = hasTimelineFailed
+            }
+        )
 
         fun handleEvent(event: PinnedMessagesBannerEvents) {
             when (event) {
                 is PinnedMessagesBannerEvents.MoveToNextPinned -> {
-                    if (currentPinnedMessageIndex < pinnedMessageCount - 1) {
-                        currentPinnedMessageIndex++
+                    if (currentPinnedMessageIndex > 0) {
+                        currentPinnedMessageIndex--
                     } else {
-                        currentPinnedMessageIndex = 0
+                        currentPinnedMessageIndex = pinnedItems.value.size - 1
```

Comment:
Not a big deal, but maybe this could be done with just `(currentPinnedMessageIndex - 1).floorMod(pinnedItems.value.size)`.